odoo/models.py

```diff
@@ -2570,55 +2575,56 @@ def _read_group_format_result_properties(self, rows_dict, group):
                 if not row[fullname]:
                     # can not only do ('many2one', '=', False) because we might have
                     # record in database that does not exist anymore
-                    additional_domain = expression.OR([
-                        [(fullname, '=', False)],
-                        [(fullname, 'not in', all_groups)],
-                    ])
+                    additional_domain = [
+                        '|',
+                        (fullname, '=', False),
+                        (fullname, 'not in', all_groups),
+                    ]
                 else:
                     additional_domain = [(fullname, '=', row[fullname])]
                     record = self.env[comodel].browse(row[fullname]).with_prefetch(prefetch_ids)
                     row[fullname] = (row[fullname], record.display_name)
 
-                row['__domain'] = expression.AND([row['__domain'], additional_domain])
+                row['__domain'] = ['&', *row['__domain'], *additional_domain]
 
         elif property_type == 'many2many':
             comodel = definition.get('comodel')
             prefetch_ids = tuple(row[fullname] for row in rows_dict if row[fullname])
             all_groups = tuple(row[fullname] for row in rows_dict if row[fullname])
             for row in rows_dict:
                 if not row[fullname]:
-                    additional_domain = expression.OR([
-                        [(fullname, '=', False)],
-                        expression.AND([[(fullname, 'not in', group)] for group in all_groups]),
-                    ]) if all_groups else []
+                    if all_groups:
+                        additional_domain = list(Domain(fullname, '=', False) | Domain.AND(*([(fullname, 'not in', group)] for group in all_groups)))
```

Comment:
The domain is rebuilt based on the values for each row in order of the group by. So for performance reasons, I try avoiding conversion between lists and domains (unless I want a normalized domain like here).
I'm updating the function so that the initial domain is extended using this list. The result is not normalized anymore, but still correct.

 @ryv-odoo I would have to either have to normalize the domains at the end of the `_read_group_format_result` function or update the tests.

Reply:
Updated by ryv. Can stay as-is or will be modified.